src/control/lib/hardware/topology.go

```diff
@@ -34,20 +41,130 @@ func (t *Topology) AllDevices() map[string]*Device {
 	return devsByName
 }
 
+func PrintTopology(t *Topology, output io.Writer) error {
+	if t == nil {
+		fmt.Fprintf(output, "No topology information available\n")
+	}
+
+	for _, numaNode := range t.NUMANodes {
+		coreSet := &system.RankSet{}
+		for _, core := range numaNode.Cores {
+			coreSet.Add(system.Rank(core.ID))
+		}
+
+		fmt.Fprintf(output, "NUMA node %d (%s)\n", numaNode.ID, numaNode.Set)
+		fmt.Fprintf(output, "  CPU cores: %s\n", coreSet)
+		fmt.Fprintf(output, "  PCI devices: \n")
+		for addr, devs := range numaNode.PCIDevices {
+			fmt.Fprintf(output, "    %s\n", &addr)
+			for _, dev := range devs {
+				fmt.Fprintf(output, "      %s\n", dev)
+			}
+		}
+		fmt.Fprintf(output, "  PCI buses: \n")
+		for _, bus := range numaNode.PCIBuses {
+			fmt.Fprintf(output, "    %s\n", bus)
+		}
+	}
+
+	return nil
+}
+
 type (
+	CPUCore struct {
+		ID       uint      `json:"id"`
+		NUMANode *NUMANode `json:"-"`
+	}
+
+	// PCIDevice represents an individual hardware device.
+	PCIDevice struct {
```

Comment:
Should probably link the PCIDevice back to its PCIBus as well.

Reply:
Yeah, was thinking about that, hadn't gotten to it yet. I think I should be able to add a method to the PCIBus type that allows it to say that the device address is on the bus, kind of like `hwloc_bridge_covers_pcibus()`